### PR TITLE
FP [6,8,12] quantizer op

### PR DESCRIPTION
**Pull request by jeffra**
Flexible-bit quantizer-dequantizer library with fp6/fp12/fp8 support

Requires Ampere+ architecture, this is due to the initial focus of this op only on `bfloat16` input types.

Co-authored-by: Reza Yazdani <reza.yazdani@snowflake.com>

### CODE REVIEW

**Comment by jeffra** [comment]
Regarding the `nv-pre-compile-ops` test failure. I just did some debugging on this and I see that for some reason if you set `DS_ENABLE_NINJA=1` this fails but if you don't use ninja the pre-compile works fine. I remember we used to have lots of issues with using ninja but then it was re-enabled in https://github.com/microsoft/DeepSpeed/pull/5088/. Not sure if this PR introduces something sublte that doesn't work w. ninja?

@mrwyattii @loadams any ideas?

**Comment by jeffra** [comment]
@microsoft-github-policy-service agree company="Snowflake"

**Comment by loadams** [comment]
> Regarding the `nv-pre-compile-ops` test failure. I just did some debugging on this and I see that for some reason if you set `DS_ENABLE_NINJA=1` this fails but if you don't use ninja the pre-compile works fine. I remember we used to have lots of issues with using ninja but then it was re-enabled in #5088. Not sure if this PR introduces something sublte that doesn't work w. ninja?
> 
> @mrwyattii @loadams any ideas?

@jeffra - No immediate guesses, but I'll take a look

**Comment by jeffra** [comment]
Thanks @loadams! I believe this is resolved now, I figured out the issue with ninja enabled. I have to disable the pre-compile for this op for the test since it includes V100 compute target which is not compatible with this op (since it only works with bf16). I was able to get all the CI tests to pass (before updating with latest master right now). 

I think now it just needs a review from folks? :)

**Comment by loadams** [comment]
> Thanks @loadams! I believe this is resolved now, I figured out the issue with ninja enabled. I have to disable the pre-compile for this op for the test since it includes V100 compute target which is not compatible with this op (since it only works with bf16). I was able to get all the CI tests to pass (before updating with latest master right now).
> 
> I think now it just needs a review from folks? :)

That makes sense, I hadn't reviewed which is probably why I missed that you already identified the issue :)

**Review by mrwyattii — APPROVED** [review]
@arashashari and/or @JamesTheZ please add your review as well, thank you

File: deepspeed/ops/fp_quantizer/quantize.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        if fp_quant_module is None:
+            fp_quant_module = FPQuantizerBuilder().load()
+
+        self.group_size = group_size
+        self.orig_dtype = None
+
+    def quantize(self,
+                 input,
+                 q_bits=8,
+                 q_mantisa_bits=3,
```
Comment (JamesTheZ) [inline deepspeed/ops/fp_quantizer/quantize.py:26]: NIT: `q_mantisa_bits` seems only used for `q_bits == 8`. This parameter can be moved and directly assign its value in L37? Or moving the predefined `q_mantisa_bits ` values in L38-43.

Comment (sfc-gh-reyazda) [reply deepspeed/ops/fp_quantizer/quantize.py:26]: that's a good point @JamesTheZ, the reason we are receiving as the parameter for FP8 is that we can switch between E5M2 and E4M3, but for the rest we hard-code the bit-ratios the kernel supports.

File: csrc/fp_quantizer/quantize.cu [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+                    dst_mantisa, dst_exponent, &state);
+                if (cur_exponent != 0)
+                    clip<_mantisa_bits, _exponent_bits, q_mantisa_bits, q_exponent_bits>(
+                        dst_exponent, dst_mantisa);
+
+                dst_mantisa = (dst_mantisa & mantisa_mask) >> (_mantisa_bits - q_mantisa_bits);
+
+                if (dst_exponent != (1 << q_exponent_bits) - 1)
+                    dst_exponent = (dst_exponent - ((1 << (_exponent_bits - 1)) - 1)) +
+                                   (1 << (q_exponent_bits - 1)) - 1;
```
[the remark this answers is not in the document]
Comment (sfc-gh-reyazda) [reply csrc/fp_quantizer/quantize.cu:144]: Yes, it is handled in the clip and round functions and also using the condition on line 136.

**Comment by sfc-gh-reyazda** [comment]
> @arashashari and/or @JamesTheZ please add your review as well, thank you

@mrwyattii, probably you meant @arashb :) btw, @arashashari is also welcome to review this ;)

**Review by JamesTheZ — APPROVED** [review]
LGTM

**Pull request merged**